### PR TITLE
Fix muxListener.Close to actually close its channel of connections

### PR DESCRIPTION
**Pull request by a-robinson**
Go through some extra pains to make it idempotent since it gets called
twice during server shutdown. It isn't exactly thread-safe as written,
though.

Fixes problem of getting stuck during shutdown in https://github.com/cockroachdb/cockroach/blob/master/pkg/server/server.go#L505 after https://github.com/cockroachdb/cockroach/pull/12743 is applied.

### CODE REVIEW

**Comment by a-robinson** [comment]
Updated to make it actually thread-safe.

**Comment by a-robinson** [comment]
@tamird

**Comment by a-robinson** [comment]
Ah, I just realized this was a fork. Should I redirect this upstream?

**Comment by tamird** [comment]
Nah, this is a long-lived fork.

I don't follow this change. My understanding of the previous code was that you'd close the underlying listener, which would result in the unblocking of the accept loop, which would close connc. What wasn't working?

**Comment by a-robinson** [comment]
Sorry for the poor explanation. What you say is correct, but only if `Serve` is ever called on the `cMux`. However, in the case when we fail to initialize the server's node (https://github.com/cockroachdb/cockroach/blob/master/pkg/server/server.go#L592), we never actually start serving on the mux (https://github.com/cockroachdb/cockroach/blob/master/pkg/server/server.go#L625) and thus the channel is never closed.

This prevents the worker that's doing grpc serving (https://github.com/cockroachdb/cockroach/blob/master/pkg/server/server.go#L505) from ever finishing, because it gets stuck waiting on the `muxListener`'s channel:
https://github.com/cockroachdb/vendored/blob/1e2620560b6dc0b7f97e0739e874cbe3386fe5a7/google.golang.org/grpc/server.go#L354
https://github.com/cockroachdb/cmux/blob/master/cmux.go#L184

Does the new commit message make a little more sense?

**Comment by tamird** [comment]
Ah, yes, this is a better explanation. Could you add a test?

File: cmux.go [hunk 1]
```diff
@@ -177,7 +178,15 @@ func (m *cMux) handleErr(err error) bool {
 
 type muxListener struct {
 	net.Listener
-	connc chan net.Conn
+	connc  chan net.Conn
+	closed *sync.Once
```
Comment (tamird) [inline cmux.go:182]: can't this be a value, btw?

Comment (a-robinson) [reply cmux.go:182]: Unfortunately it can't because the muxListener gets passed around by value, and copying a sync.Once object around is a no-no.

**Comment by a-robinson** [comment]
Test added. Before the fix it hangs forever, after the fix it passes trivially.

**Comment by tamird** [comment]
LGTM 

Feel free to send this upstream as well.

**Comment by tamird** [comment]
This seems to be unsafe - if `Close()` is called externally concurrent with a send on the channel...boom.

**Comment by a-robinson** [comment]
Boom indeed... I clearly didn't understand this fully before fixing it. Taking a closer look...

**Pull request merged**